File: windows/BlockedWindow.cs
using System.Drawing;

namespace Insight;

/// <summary>
/// What replaces a blocked app: a small window that says what happened.
///
/// The app has been closed by the time this appears, so "Let me in anyway"
/// starts it again rather than merely restoring a minimised window.
///
/// The override is deliberately available and deliberately slightly slow. The
/// plan calls for a soft commitment rather than a hard lock, because a hard
/// lock gets the app uninstalled, and an uninstalled app records nothing at
/// all. Three seconds interrupts the reflex without becoming a punishment.
///
/// Laid out in panels rather than at pixel coordinates, for the reason given
/// in <see cref="PairWindow"/>: on a display-scaled laptop, fixed coordinates
/// put the text on top of the buttons.
/// </summary>
internal sealed class BlockedWindow : Form
{
    private const int ContentWidth = 400;

    private readonly string _app;
    private readonly Tracker _tracker;
    private readonly Button _override;
    private readonly System.Windows.Forms.Timer _countdown;
    private int _remaining = 3;

    internal BlockedWindow(string app, Tracker tracker)
    {
        _app = app;
        _tracker = tracker;

        Text = "Insight";
        BackColor = Theme.Background;
        ForeColor = Theme.Text;
        Font = Theme.Body;
        FormBorderStyle = FormBorderStyle.FixedDialog;
        MaximizeBox = false;
        MinimizeBox = false;
        ShowInTaskbar = false;
        TopMost = true;
        StartPosition = FormStartPosition.CenterScreen;
        AutoScaleMode = AutoScaleMode.Font;

        var layout = new TableLayoutPanel
        {
            ColumnCount = 1,
            AutoSize = true,
            AutoSizeMode = AutoSizeMode.GrowAndShrink,
            Dock = DockStyle.Fill,
            Padding = new Padding(24, 22, 24, 22),
            BackColor = Theme.Background,
        };
        layout.ColumnStyles.Add(new ColumnStyle(SizeType.AutoSize));

        var heading = new Label
        {
            Text = "Focus Mode is on",
            Font = Theme.Heading,
            ForeColor = Theme.Text,
            AutoSize = true,
            Margin = new Padding(0, 0, 0, 8),
        };

        var body = new Label
        {
            Text = $"{app} is on your blocklist, so it's been closed. If it had "
                 + "unsaved work it will have asked you first.",
            ForeColor = Theme.TextMuted,
            AutoSize = true,
            MaximumSize = new Size(ContentWidth, 0),
            Margin = new Padding(0, 0, 0, 20),
        };

        var back = new Button
        {
            Text = "Back to work",
            AutoSize = true,
            AutoSizeMode = AutoSizeMode.GrowAndShrink,
            Padding = new Padding(18, 8, 18, 8),
            BackColor = Theme.Accent,
            ForeColor = Color.Black,
            FlatStyle = FlatStyle.Flat,
            Margin = new Padding(0, 0, 12, 0),
        };
        back.FlatAppearance.BorderSize = 0;
        back.Click += (_, _) => Close();

        _override = new Button
        {
            Text = "Let me in anyway",
            AutoSize = true,
            AutoSizeMode = AutoSizeMode.GrowAndShrink,
            Padding = new Padding(18, 8, 18, 8),
            BackColor = Theme.Surface,
            ForeColor = Theme.TextMuted,
            FlatStyle = FlatStyle.Flat,
            Margin = new Padding(0),
        };
        _override.FlatAppearance.BorderColor = Theme.Line;
        _override.Click += StartOverride;

        var buttons = new FlowLayoutPanel
        {
            AutoSize = true,
            AutoSizeMode = AutoSizeMode.GrowAndShrink,
            FlowDirection = FlowDirection.LeftToRight,
            WrapContents = false,
            Margin = new Padding(0, 0, 0, 16),
            Padding = new Padding(0),
        };
        buttons.Controls.Add(back);
        buttons.Controls.Add(_override);

        var footnote = new Label
        {
            Text = "Overrides are recorded, so your figures reflect what actually happened.",
            ForeColor = Theme.TextFaint,
            Font = Theme.Small,
            AutoSize = true,
            MaximumSize = new Size(ContentWidth, 0),
            Margin = new Padding(0),
        };

        layout.Controls.Add(heading);
        layout.Controls.Add(body);
        layout.Controls.Add(buttons);
        layout.Controls.Add(footnote);

        Controls.Add(layout);
        ClientSize = layout.PreferredSize;

        AcceptButton = back;

        _countdown = new System.Windows.Forms.Timer { Interval = 1000 };
        _countdown.Tick += OnCountdownTick;
    }

    private void StartOverride(object? sender, EventArgs e)
    {
        if (_countdown.Enabled) return;

        _override.Text = $"Opening in {_remaining}…";
        _countdown.Start();
    }

    private void OnCountdownTick(object? sender, EventArgs e)
    {
        _remaining--;
        if (_remaining > 0)
        {
            _override.Text = $"Opening in {_remaining}…";
            return;
        }

        _countdown.Stop();

        // Recorded before it takes effect, so a session that ends mid-override
        // still shows the override.
        _tracker.RecordOverride(_app);
        Close();
    }

    protected override void OnFormClosed(FormClosedEventArgs e)
    {
        _countdown.Dispose();
        base.OnFormClosed(e);
    }
}

File: mac/Sources/Insight/BlockedWindow.swift
import AppKit

/// What a blocked app is replaced with: a small window that says what
/// happened.
///
/// The app has been quit by the time this appears, so "Let me in anyway"
/// starts it again rather than merely un-hiding it.
///
/// The override is deliberately available and deliberately slightly slow. The
/// plan calls for a soft commitment rather than a hard lock, because a hard
/// lock gets the app uninstalled, and an uninstalled app records nothing at
/// all. Three seconds interrupts the reflex without becoming a punishment.
@MainActor
final class BlockedWindow: NSWindowController {
    private static let contentWidth: CGFloat = 400

    private let app: String
    private let tracker: Tracker
    private var overrideButton: NSButton!
    private var countdown: Timer?
    private var remaining = 3

    init(app: String, tracker: Tracker) {
        self.app = app
        self.tracker = tracker

        let window = NSWindow(
            contentRect: NSRect(x: 0, y: 0, width: 448, height: 220),
            styleMask: [.titled, .closable],
            backing: .buffered,
            defer: false)
        window.title = "Insight"
        window.appearance = NSAppearance(named: .darkAqua)
        window.backgroundColor = Theme.background
        window.level = .floating
        window.isReleasedWhenClosed = false

        super.init(window: window)

        let back = Views.button("Back to work", primary: true) { [weak self] in
            self?.close()
        }

        overrideButton = Views.button("Let me in anyway", primary: false) { [weak self] in
            self?.startOverride()
        }

        let buttons = NSStackView(views: [back, overrideButton])
        buttons.orientation = .horizontal
        buttons.spacing = 12

        let stack = Views.column()
        stack.setViews([
            Views.heading("Focus Mode is on"),
            Views.prose(
                "\(app) is on your blocklist, so it's been closed. If it had unsaved "
                + "work it will have asked you first."),
            buttons,
            Views.footnote(
                "Overrides are recorded, so your figures reflect what actually happened."),
        ], in: .leading)

        stack.setCustomSpacing(20, after: stack.views[1])
        stack.setCustomSpacing(16, after: buttons)

        guard let contentView = window.contentView else { return }
        contentView.addSubview(stack)

        NSLayoutConstraint.activate([
            stack.topAnchor.constraint(equalTo: contentView.topAnchor),
            stack.leadingAnchor.constraint(equalTo: contentView.leadingAnchor),
            stack.trailingAnchor.constraint(equalTo: contentView.trailingAnchor),
            stack.bottomAnchor.constraint(lessThanOrEqualTo: contentView.bottomAnchor),
        ])

        for view in stack.views {
            if let field = view as? NSTextField {
                field.preferredMaxLayoutWidth = BlockedWindow.contentWidth
            }
        }

        window.center()
    }

    required init?(coder: NSCoder) {
        fatalError("not used")
    }

    func present() {
        showWindow(nil)
        window?.makeKeyAndOrderFront(nil)
        NSApp.activate(ignoringOtherApps: true)
    }

    private func startOverride() {
        guard countdown == nil else { return }

        overrideButton.title = "Opening in \(remaining)…"

        let timer = Timer(timeInterval: 1, repeats: true) { [weak self] _ in
            Task { @MainActor in self?.step() }
        }
        RunLoop.main.add(timer, forMode: .common)
        countdown = timer
    }

    private func step() {
        remaining -= 1
        if remaining > 0 {
            overrideButton.title = "Opening in \(remaining)…"
            return
        }

        countdown?.invalidate()
        countdown = nil

        // Recorded before it takes effect, so a session that ends mid-override
        // still shows the override.
        tracker.recordOverride(app)
        close()
    }

    override func close() {
        countdown?.invalidate()
        countdown = nil
        super.close()
    }
}
